pkg/dns/google/google.go

```diff
@@ -0,0 +1,500 @@
+
+	return changes
+}
+
+// getResourceRecordSets returns the records for a managed zone of the currently configured provider.
+func (g *GoogleDNSProvider) getResourceRecordSets(ctx context.Context, zoneID string) ([]*dnsv1.ResourceRecordSet, error) {
+	var records []*dnsv1.ResourceRecordSet
+
+	f := func(resp *dnsv1.ResourceRecordSetsListResponse) error {
+		records = append(records, resp.Rrsets...)
+		return nil
+	}
+
+	if err := g.resourceRecordSetsClient.List(g.project, zoneID).Pages(ctx, f); err != nil {
+		return nil, err
+	}
+
+	return records, nil
+}
+
+// toResourceRecordSets converts a list of endpoints into `ResourceRecordSet` resources.
+func toResourceRecordSets(allEndpoints []*v1alpha1.Endpoint) []*dnsv1.ResourceRecordSet {
+	var records []*dnsv1.ResourceRecordSet
+
+	// Google DNS requires a record to be created per `dnsName`, so the first thing we need to do is group all the
+	// endpoints with the same dnsName together.
+	endpointMap := make(map[string][]*v1alpha1.Endpoint)
+	for _, ep := range allEndpoints {
+		endpointMap[ep.DNSName] = append(endpointMap[ep.DNSName], ep)
+	}
+
+	for dnsName, endpoints := range endpointMap {
+		// A set of endpoints belonging to the same group(`dnsName`) must always be of the same type, have the same ttl
+		// and contain the same rrdata (weighted or geo), so we can just get that from the first endpoint in the list.
+		ttl := int64(endpoints[0].RecordTTL)
+		recordType := endpoints[0].RecordType
+		_, weighted := endpoints[0].GetProviderSpecificProperty(dns.ProviderSpecificWeight)
+		_, geoCode := endpoints[0].GetProviderSpecificProperty(dns.ProviderSpecificGeoCode)
+
+		record := &dnsv1.ResourceRecordSet{
+			Name: ensureTrailingDot(dnsName),
+			Ttl:  ttl,
+			Type: recordType,
+		}
+		if weighted {
+			record.RoutingPolicy = &dnsv1.RRSetRoutingPolicy{
+				Wrr: &dnsv1.RRSetRoutingPolicyWrrPolicy{},
+			}
+		} else if geoCode {
+			record.RoutingPolicy = &dnsv1.RRSetRoutingPolicy{
+				Geo: &dnsv1.RRSetRoutingPolicyGeoPolicy{},
+			}
+		}
+
+		for _, ep := range endpoints {
+			targets := make([]string, len(ep.Targets))
+			copy(targets, ep.Targets)
+			if ep.RecordType == string(v1alpha1.CNAMERecordType) {
+				targets[0] = ensureTrailingDot(targets[0])
+			}
+
+			if !weighted && !geoCode {
+				record.Rrdatas = targets
+			}
+			if weighted {
+				weightProp, _ := ep.GetProviderSpecificProperty(dns.ProviderSpecificWeight)
+				weight, err := strconv.ParseFloat(weightProp.Value, 64)
+				if err != nil {
+					weight = 0
+				}
+				item := &dnsv1.RRSetRoutingPolicyWrrPolicyWrrPolicyItem{
+					Rrdatas: targets,
+					Weight:  weight,
+				}
+				record.RoutingPolicy.Wrr.Items = append(record.RoutingPolicy.Wrr.Items, item)
+			}
+			if geoCode {
+				geoCodeProp, _ := ep.GetProviderSpecificProperty(dns.ProviderSpecificGeoCode)
+				targetIsDefaultGroup := strings.HasPrefix(ep.Targets[0], string(dns.DefaultGeo))
+				if geoCodeProp.Value == "*" && !targetIsDefaultGroup {
+					continue
+				}
```

Comment:
You should probably add this [comment](https://github.com/Kuadrant/multicluster-gateway-controller/commit/55ff95795d94099bca4bc46c1a63bc83af829dbd#diff-eea59a6fb9851390f6210a03f57ba88ea1628c32d20d526998600135a1642bedR479-R481), or something similar in your own words, explaining why we are having to do this as it's probably not immediately obvious.